File: justime_agent/src/types/taskProcess.ts
export type TaskStatus = 'draft' | 'planned' | 'active' | 'paused' | 'blocked' | 'completed' | 'archived'
export type TaskPhase = 'before' | 'during' | 'after'
export type TaskPriority = 'low' | 'medium' | 'high' | 'critical'
export type TaskCategory = 'learning' | 'development' | 'writing' | 'research' | 'reading' | 'project' | 'practice' | 'other'
export type ProgressSource = 'manual' | 'ai' | 'evidence'
export type EvidenceType = 'note' | 'file' | 'chat' | 'code_commit' | 'link' | 'time_log' | 'quiz_result' | 'review' | 'milestone_complete' | 'manual'
export type KnowledgeFormat = 'summary' | 'tutorial' | 'faq' | 'cheatsheet' | 'debug_log' | 'mindmap' | 'glossary' | 'timeline' | 'comparison' | 'review' | 'notes'

export interface Milestone {
  id: string
  title: string
  description: string
  order: number
  status: 'pending' | 'active' | 'completed' | 'skipped'
  target_date?: string | null
  completed_at?: string | null
}

export interface LearningMaterial {
  title: string
  url?: string | null
  summary: string
  source: string
}

export interface PreparationItem {
  id: string
  title: string
  done: boolean
  order: number
}

export interface Blocker {
  id: string
  description: string
  severity: 'low' | 'medium' | 'high'
  resolved: boolean
  resolution?: string | null
  created_at: string
  resolved_at?: string | null
}

export interface AISuggestion {
  id: string
  type: 'next_step' | 'resource' | 'review' | 'alert' | 'optimization'
  content: string
  accepted?: boolean | null
  created_at: string
}

export interface AIAssessment {
  progress: number
  confidence: number
  summary: string
  blockers_identified: string[]
  next_steps: string[]
  assessed_at: string
}

export interface TaskProcess {
  id: string
  userId: string
  title: string
  description: string
  goal: string
  category: TaskCategory
  tags: string[]
  status: TaskStatus
  phase: TaskPhase
  priority: TaskPriority
  progress: number
  progress_source: ProgressSource
  estimated_hours?: number | null
  actual_hours: number
  started_at?: string | null
  completed_at?: string | null
  deadline?: string | null
  materials?: LearningMaterial[]
  preparation_items?: PreparationItem[]
  milestones: Milestone[]
  blockers: Blocker[]
  ai_suggestions: AISuggestion[]
  ai_last_assessment?: AIAssessment | null
  ai_plan?: Record<string, unknown> | null
  parent_task_id?: string | null
  related_chat_session_ids: string[]
  related_calendar_event_ids: string[]
  evidence_count: number
  knowledge_output_count: number
  createdAt?: string | null
  updatedAt?: string | null
}

export interface Evidence {
  id: string
  task_id: string
  userId: string
  type: EvidenceType
  title: string
  content: string
  source: string
  milestone_id?: string | null
  metadata?: Record<string, unknown> | null
  ai_extracted: boolean
  sentiment?: 'positive' | 'neutral' | 'negative' | 'blocked' | null
  confidence: number
  createdAt?: string | null
  updatedAt?: string | null
}

export interface KnowledgeOutput {
  id: string
  task_id: string
  userId: string
  title: string
  format: KnowledgeFormat
  markdown: string
  vault_relative_path: string
  obsidian_tags: string[]
  obsidian_links: string[]
  status: 'draft' | 'reviewed' | 'published' | 'archived'
  source_evidence_ids: string[]
  absolute_path?: string | null
  published_at?: string | null
  word_count: number
  version: number
  previous_version_id?: string | null
  version_history?: KnowledgeOutputVersion[]
  createdAt?: string | null
  updatedAt?: string | null
}

export interface KnowledgeOutputVersion {
  version: number
  title: string
  markdown: string
  vault_relative_path: string
  status: KnowledgeOutput['status']
  updated_at: string
  published_at?: string | null
}

export interface VaultConfig {
  vault_root_path: string
  auto_publish: boolean
  default_category_mapping: Record<string, string>
  frontmatter_template?: Record<string, unknown>
}

export interface TaskProcessCreatePayload {
  title: string
  description: string
  goal: string
  category: TaskCategory
  tags: string[]
  priority: TaskPriority
  estimated_hours?: number | null
  deadline?: string | null
  auto_plan?: boolean
}

export interface TaskProcessUpdatePayload {
  title?: string
  description?: string
  goal?: string
  category?: TaskCategory
  tags?: string[]
  status?: TaskStatus
  phase?: TaskPhase
  priority?: TaskPriority
  progress?: number
  estimated_hours?: number | null
  deadline?: string | null
  preparation_items?: PreparationItem[]
}

export interface EvidenceCreatePayload {
  task_id: string
  type: EvidenceType
  title: string
  content: string
  source: string
  milestone_id?: string | null
  metadata?: Record<string, unknown> | null
}

export interface TimeLogCreatePayload {
  task_id: string
  hours: number
  date?: string | null
  notes: string
  milestone_id?: string | null
}

export interface GenerateKnowledgePayload {
  task_id: string
  format: KnowledgeFormat
  additional_instructions?: string
  include_evidence_ids?: string[]
}

export interface KnowledgeOutputUpdatePayload {
  title?: string
  markdown?: string
  format?: KnowledgeFormat
  status?: KnowledgeOutput['status']
  vault_relative_path?: string
  obsidian_tags?: string[]
  obsidian_links?: string[]
}

export interface KnowledgeRollbackPayload {
  version: number
}

File: justime_backend/app/models/task_process.py
"""
TaskProcess data models.

TaskProcess is the core lifecycle entity for Justime:
before -> during -> after.
"""

import re
from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


TaskStatus = Literal[
    "draft",
    "planned",
    "active",
    "paused",
    "blocked",
    "completed",
    "archived",
]

TaskPhase = Literal["before", "during", "after"]
TaskPriority = Literal["low", "medium", "high", "critical"]
TaskCategory = Literal[
    "learning",
    "development",
    "writing",
    "research",
    "reading",
    "project",
    "practice",
    "other",
]
ProgressSource = Literal["manual", "ai", "evidence"]
MilestoneStatus = Literal["pending", "active", "completed", "skipped"]
AgentMode = Literal["plan", "research", "monitor", "coach", "summarize", "knowledge"]


def _validate_loose_id(value: str, field_name: str) -> str:
    value = (value or "").strip()
    if not value:
        raise ValueError(f"{field_name}不能为空")
    if not re.match(r"^[a-zA-Z0-9_-]+$", value):
        raise ValueError(f"{field_name}格式无效")
    return value


class Milestone(BaseModel):
    id: str = Field(..., max_length=50)
    title: str = Field(..., min_length=1, max_length=200)
    description: str = Field("", max_length=2000)
    order: int = Field(0, ge=0)
    status: MilestoneStatus = "pending"
    target_date: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    @field_validator("id")
    @classmethod
    def validate_id(cls, value: str) -> str:
        return _validate_loose_id(value, "里程碑ID")

    @field_validator("title")
    @classmethod
    def validate_title(cls, value: str) -> str:
        value = (value or "").strip()
        if not value:
            raise ValueError("里程碑标题不能为空")
        return value


class MilestoneStatusUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: MilestoneStatus


class LearningMaterial(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    url: Optional[str] = Field(None, max_length=2000)
    summary: str = Field(..., min_length=1, max_length=2000)
    source: str = Field(..., min_length=1, max_length=200)

    @field_validator("title", "summary", "source")
    @classmethod
    def validate_required_text(cls, value: str) -> str:
        value = (value or "").strip()
        if not value:
            raise ValueError("学习资料字段不能为空")
        return value

    @field_validator("url")
    @classmethod
    def validate_url(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return value
        return value.strip() or None


class PreparationItem(BaseModel):
    id: str = Field(..., max_length=50)
    title: str = Field(..., min_length=1, max_length=200)
    done: bool = False
    order: int = Field(0, ge=0)

    @field_validator("id")
    @classmethod
    def validate_id(cls, value: str) -> str:
        return _validate_loose_id(value, "准备项ID")

    @field_validator("title")
    @classmethod
    def validate_title(cls, value: str) -> str:
        value = (value or "").strip()
        if not value:
            raise ValueError("准备项标题不能为空")
        return value


class Blocker(BaseModel):
    id: str = Field(..., max_length=50)
    description: str = Field(..., min_length=1, max_length=2000)
    severity: Literal["low", "medium", "high"] = "medium"
    resolved: bool = False
    resolution: Optional[str] = Field(None, max_length=2000)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    resolved_at: Optional[datetime] = None

    @field_validator("id")
    @classmethod
    def validate_id(cls, value: str) -> str:
        return _validate_loose_id(value, "阻塞点ID")

    @field_validator("description")
    @classmethod
    def validate_description(cls, value: str) -> str:
        value = (value or "").strip()
        if not value:
            raise ValueError("阻塞描述不能为空")
        return value


class AISuggestion(BaseModel):
    id: str = Field(..., max_length=50)
    type: Literal["next_step", "resource", "review", "alert", "optimization"]
    content: str = Field(..., min_length=1, max_length=5000)
    accepted: Optional[bool] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)

    @field_validator("id")
    @classmethod
    def validate_id(cls, value: str) -> str:
        return _validate_loose_id(value, "建议ID")


class AIAssessment(BaseModel):
    progress: float = Field(..., ge=0.0, le=1.0)
    confidence: float = Field(..., ge=0.0, le=1.0)
    summary: str = Field("", max_length=2000)
    blockers_identified: List[str] = Field(default_factory=list)
    next_steps: List[str] = Field(default_factory=list)
    assessed_at: datetime = Field(default_factory=datetime.utcnow)


class TaskProcessBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: str = Field("", max_length=5000)
    goal: str = Field(..., min_length=1, max_length=2000)
    category: TaskCategory = "other"
    tags: List[str] = Field(default_factory=list)
    priority: TaskPriority = "medium"
    estimated_hours: Optional[float] = Field(None, ge=0.0, le=10000)
    deadline: Optional[datetime] = None
    materials: List[LearningMaterial] = Field(default_factory=list)
    preparation_items: List[PreparationItem] = Field(default_factory=list)

    @field_validator("title", "goal")
    @classmethod
    def validate_required_text(cls, value: str) -> str:
        value = (value or "").strip()
        if not value:
            raise ValueError("必填字段不能为空")
        return value

    @field_validator("description")
    @classmethod
    def validate_description(cls, value: str) -> str:
        return (value or "").strip()

    @field_validator("tags")
    @classmethod
    def validate_tags(cls, value: List[str]) -> List[str]:
        cleaned = [tag.strip() for tag in (value or []) if tag and tag.strip()]
        if len(cleaned) > 20:
            raise ValueError("标签数量不能超过20个")
        return cleaned


class TaskProcessCreate(TaskProcessBase):
    initial_chat_session_id: Optional[str] = Field(None, max_length=100)
    initial_calendar_event_id: Optional[str] = Field(None, max_length=100)
    auto_plan: bool = True


class TaskProcessUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=5000)
    goal: Optional[str] = Field(None, min_length=1, max_length=2000)
    category: Optional[TaskCategory] = None
    tags: Optional[List[str]] = None
    status: Optional[TaskStatus] = None
    phase: Optional[TaskPhase] = None
    priority: Optional[TaskPriority] = None
    progress: Optional[float] = Field(None, ge=0.0, le=1.0)
    estimated_hours: Optional[float] = Field(None, ge=0.0, le=10000)
    deadline: Optional[datetime] = None
    materials: List[LearningMaterial] = Field(default_factory=list)
    preparation_items: List[PreparationItem] = Field(default_factory=list)

    @field_validator("title", "goal", "description")
    @classmethod
    def validate_optional_text(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return value
        value = value.strip()
        if not value:
            raise ValueError("字段不能为空字符串")
        return value

    @field_validator("tags")
    @classmethod
    def validate_tags(cls, value: Optional[List[str]]) -> Optional[List[str]]:
        if value is None:
            return value
        cleaned = [tag.strip() for tag in value if tag and tag.strip()]
        if len(cleaned) > 20:
            raise ValueError("标签数量不能超过20个")
        return cleaned


class TaskProcessOut(TaskProcessBase):
    id: str
    userId: str
    status: TaskStatus = "draft"
    phase: TaskPhase = "before"
    progress: float = Field(0.0, ge=0.0, le=1.0)
    progress_source: ProgressSource = "manual"
    actual_hours: float = 0.0
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    milestones: List[Milestone] = Field(default_factory=list)
    blockers: List[Blocker] = Field(default_factory=list)
    ai_suggestions: List[AISuggestion] = Field(default_factory=list)
    ai_last_assessment: Optional[AIAssessment] = None
    ai_plan: Optional[Dict[str, Any]] = None
    parent_task_id: Optional[str] = None
    related_chat_session_ids: List[str] = Field(default_factory=list)
    related_calendar_event_ids: List[str] = Field(default_factory=list)
    evidence_count: int = 0
    knowledge_output_count: int = 0
    createdAt: Optional[datetime] = None
    updatedAt: Optional[datetime] = None


class MilestoneCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: str = Field("", max_length=2000)
    target_date: Optional[datetime] = None
    order: int = Field(0, ge=0)

    @field_validator("title")
    @classmethod
    def validate_title(cls, value: str) -> str:
        value = (value or "").strip()
        if not value:
            raise ValueError("里程碑标题不能为空")
        return value


class MilestoneUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    status: Optional[MilestoneStatus] = None
    target_date: Optional[datetime] = None
    order: Optional[int] = Field(None, ge=0)


class BlockerCreate(BaseModel):
    description: str = Field(..., min_length=1, max_length=2000)
    severity: Literal["low", "medium", "high"] = "medium"

    @field_validator("description")
    @classmethod
    def validate_description(cls, value: str) -> str:
        value = (value or "").strip()
        if not value:
            raise ValueError("阻塞描述不能为空")
        return value


class BlockerResolve(BaseModel):
    resolution: str = Field(..., min_length=1, max_length=2000)

    @field_validator("resolution")
    @classmethod
    def validate_resolution(cls, value: str) -> str:
        value = (value or "").strip()
        if not value:
            raise ValueError("解决方案不能为空")
        return value


class TaskAgentRequest(BaseModel):
    task_id: str = Field(..., max_length=100)
    mode: AgentMode
    user_input: str = Field("", max_length=10000)
    context: Optional[Dict[str, Any]] = None

    @field_validator("task_id")
    @classmethod
    def validate_task_id(cls, value: str) -> str:
        return _validate_loose_id(value, "任务ID")


class TaskAgentResponse(BaseModel):
    success: bool
    mode: AgentMode
    result: Optional[Dict[str, Any]] = None
    suggestions: List[AISuggestion] = Field(default_factory=list)
    assessment: Optional[AIAssessment] = None
    error: Optional[str] = None


class TaskProcessListQuery(BaseModel):
    status: Optional[TaskStatus] = None
    phase: Optional[TaskPhase] = None
    category: Optional[TaskCategory] = None
    priority: Optional[TaskPriority] = None
    search: Optional[str] = Field(None, max_length=200)
    sort_by: Literal["createdAt", "updatedAt", "deadline", "priority", "progress"] = "updatedAt"
    sort_order: Literal["asc", "desc"] = "desc"
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)
